### PR TITLE
Ctcp 2796

### PR DESCRIPTION
**Pull request by srinivasulumunagalahmrc**

### CODE REVIEW

File: app/controllers/actions/RejectionMessageAction.scala [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+  protected val executionContext: ExecutionContext
+) extends ActionRefiner[IdentifierRequest, RejectionMessageRequest] {
+
+  override protected def refine[A](request: IdentifierRequest[A]): Future[Either[Result, RejectionMessageRequest[A]]] = {
+
+    implicit val hc: HeaderCarrier = HeaderCarrierConverter.fromRequestAndSession(request, request.session)
+
+    (for {
+      ie056 <- OptionT(departureP5MessageService.getRejectionMessage(departureId))
+      lrn    = ie056.data.TransitOperation.LRN.getOrElse("")
```
Comment (jackheslop96) [inline app/controllers/actions/RejectionMessageAction.scala:50]: I have a feeling we can only guarantee the presence of an LRN in an IE015. In an IE056 it's dependent on a couple of things - https://developer.service.hmrc.gov.uk/guides/ctc-traders-phase5-tis/documentation/messagetypes.html#ie056-rejection-from-office-of-departure

File: app/connectors/DepartureMovementP5Connector.scala [hunk 2]
```diff
@@ -85,4 +85,12 @@ class DepartureMovementP5Connector @Inject() (config: FrontendAppConfig, http: H
...
     http.GET[IE060Data](serviceUrl)(implicitly, headers, ec)
   }
+
+  def getRejectionMessage(path: String)(implicit ec: ExecutionContext, hc: HeaderCarrier): Future[IE056Data] = {
+    val headers = hc.withExtraHeaders(("Accept", "application/vnd.hmrc.2.0+json"))
+
+    val serviceUrl = s"${config.commonTransitConventionTradersUrl}$path"
+
+    http.GET[IE056Data](serviceUrl)(implicitly, headers, ec)
+  }
```
Comment (MatthewAlexanderBeck) [inline app/connectors/DepartureMovementP5Connector.scala:95]: This looks the same as the method above, can we maybe bring in an ambigious type to simplify e.g.

def getMessage[A](path: String)(implicit ec: ExecutionContext, hc: HeaderCarrier, reads: Reads[A]): Future[A] = ???

File: app/connectors/ReferenceDataConnector.scala [hunk 3]
```diff
@@ -51,4 +52,18 @@ class ReferenceDataConnector @Inject() (config: FrontendAppConfig, http: HttpCli
...
+
+    http
+      .GET[Option[FunctionalErrorWithDesc]](serviceUrl)
+      .map(_.getOrElse(onFailFunctionalError(code)))
+      .recover {
+        case _ =>
+          logger.error(s"Get Functional Error Type  request failed to return data")
+          onFailFunctionalError(code)
+      }
+  }
```
Comment (MatthewAlexanderBeck) [inline app/connectors/ReferenceDataConnector.scala:68]: Is this not the same as:

  http
      .GET [FunctionalErrorWithDesc] (serviceUrl)
      .recover {
        case _ =>
          logger.error(s"Get Functional Error Type  request failed to return data")
          onFailFunctionalError(code)
      }
  }

File: app/controllers/actions/RejectionMessageAction.scala [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  protected val executionContext: ExecutionContext
+) extends ActionRefiner[IdentifierRequest, RejectionMessageRequest] {
+
+  override protected def refine[A](request: IdentifierRequest[A]): Future[Either[Result, RejectionMessageRequest[A]]] = {
+
+    implicit val hc: HeaderCarrier = HeaderCarrierConverter.fromRequestAndSession(request, request.session)
+
+    (for {
+      ie056 <- OptionT(departureP5MessageService.getRejectionMessage(departureId))
+      lrn   <- OptionT(departureP5MessageService.getLRNFromDeclarationMessage(departureId))
```
Comment (MatthewAlexanderBeck) [inline app/controllers/actions/RejectionMessageAction.scala:50]: Might be worth leaving a TODO that we can remove the LRN retrieval when this ticket is done: https://jira.tools.tax.service.gov.uk/browse/CTCP-2924

File: conf/application.conf [hunk 5]
```diff
@@ -127,13 +127,20 @@ microservice {
...
       startUrl = "test-only/transit-movements-trader-reference-data"
     }
 
+    manage-transit-movements-departure-cache {
+      protocol = http
+      host = localhost
+      port = 10126
+      startUrl = "manage-transit-movements-departure-cache"
+    }
+
```
Comment (MatthewAlexanderBeck) [inline conf/application.conf:143]: Might need to add to app-config-base

File: app/connectors/ReferenceDataConnector.scala (hunk 3 above)
Comment (srinivasulumunagalahmrc) [reply app/connectors/ReferenceDataConnector.scala:68]: when the call returns None, we want to pass a default FunctionalError. so we put getOrElse. :)

**Review by MatthewAlexanderBeck — APPROVED** [review]

**Review by MatthewAlexanderBeck — DISMISSED** [review]

**Review by Reece-Carruthers — APPROVED** [review]

**Pull request merged**